### ISSUE TITLE
Unicode symbols >= 𐀀 are broken

### ISSUE THREAD

**Issue by Bodigrim** [haskell/haskell-language-server#2646]
### Your environment

Which OS do you use: MacOS
Which LSP client (editor/plugin) do you use: Sublime Code

### Steps to reproduce
```
aaa :: Word 
aaa = 10

𐀀𐀀𐀀 :: Word 
𐀀𐀀𐀀 = 10 
```

### Expected behaviour

Both `aaa` and `𐀀𐀀𐀀` should be underlined as unused bindings. 

### Actual behaviour

`aaa` is underlined correctly, but only the first character of `𐀀𐀀𐀀` is underlined.

<img width="96" alt="изображение" src="https://user-images.githubusercontent.com/2293557/151464453-6749fcbd-18c6-4cdb-b242-696430c50fe9.png">

That's because HLS does not distinguish positions returned from GHC (which are in code points = characters) from positions mandated by LSP (which are UTF-16 code units). Basically, GHC says us that 3 first code points in line 5 are an unused binding. Each `𐀀` is a single character, but 2 UTF-16 code units, so HLS should ask LSP to underline first 6 code units. Instead of this HLS asks LSP to underline only 3, and since the 3rd one is in the middle of the 2nd character, only the 1st character gets underlined.

CC @michaelpj @alanz, this is related to https://github.com/haskell/lsp/pull/392#discussion_r790164996.

**Comment by Bodigrim** [comment]
Code actions are also broken. E. g., choosing "Delete '𐀀𐀀𐀀'" results in the following program
```
aaa :: Word 
aaa = 10 10
```
That's again because HLS meant to ask LSP to delete 8 characters, but asked to delete 8 code points, 6 of which were `𐀀𐀀𐀀`.

**Comment by michaelpj** [comment]
Thanks for the reproduction. To be clear, did you mean that this is related to the issue with splitting in the middle of code units, or is it this one? https://github.com/haskell/lsp/pull/392#discussion_r785379400

That is, do you think this will be fixed by the changes you made to `lsp`, or is there still something broken?

**Comment by jneira** [comment]
In any case a regression test (or several of them) here exercising unicode symbols would be great

**Comment by Bodigrim** [comment]
@michaelpj This is unlikely to be fixed in recent `lsp` patches. Essentially there must be a place, responsible for (fast) conversion between positions in code points and positions in code units. I think I can provide such low-level routine from `text-rope` and then someone will have to migrate `ghcide` to it. 

@jneira I’m sorry, I’m unlikely to have capacity to add a regression test. Could someone else possibly please pick it up?

**Comment by michaelpj** [comment]
I'll look into it (inc tests), I just thought you might already have a hunch as to where the problem lies.

**Comment by michaelpj** [comment]
Looking at this more, I think this is utterly broken throughout the codebase. We frequently change from GHC `SrcSpan`s to `Position`s, blindly assuming that the numbers can just be translated between them, which is very wrong. And we use the numbers from `Position` with functions from `Text`, which is *also* wrong.

If I understand correctly, you can't even convert from a postion-in-code-points to a position-in-code-units without having the whole text in question to hand, which seems quite annoying. In that case it probably would be useful to have such a function in `text-rope` and we'll just have to painfully audit HLS for direct use of `SrcSpan`s.

(Maybe `text-rope` should allow using `Position`s defined in terms of code-points also, might be more useful for clients other than LSP...)

**Comment by michaelpj** [comment]
Not really sure what the best way to tackle this systematically is. We could have a `CodeUnit` newtype in `lsp` that we use for the column offset, that might force us to look at all the use sites. I guess I can't argue that the `Text` methods should take a `CodePoint` newtype instead of `Int` :)

**Comment by michaelpj** [comment]
Vague plan:
- Since we need the text, let's ask the VFS to do the conversion for us
- Add `CodePointPosition` to the VFS, which is a mirror of `Position` with code points for the column offset. Perhaps also `CodePointRange`, `CodePointLocation`
- Add a function of type `VFS -> URI -> CodePointPosition -> Position`, etc.

**Comment by Bodigrim** [comment]
> (Maybe `text-rope` should allow using `Position`s defined in terms of code-points also, might be more useful for clients other than LSP...)

Already there: it provides both character-based API (`Data.Text.Lines` / `Data.Text.Rope`) and UTF16-based API (`Data.Text.Utf16.Lines` / `Data.Text.Utf16.Rope`). I just need to provide conversions.

**Comment by Bodigrim** [comment]
I realised that `text-rope` already provides a way to convert between positions: one can [`Data.Text.Lines.splitAtPosition`](https://hackage.haskell.org/package/text-rope-0.1/docs/Data-Text-Lines.html#v:splitAtPosition) with character-based counter and then call [`Data.Text.Utf16.Lines.lengthAsPosition`](https://hackage.haskell.org/package/text-rope-0.1/docs/Data-Text-Utf16-Lines.html#v:lengthAsPosition) on the prefix to receive UTF-16-based counter. I might provide a dedicated helper in the future, but this would work for a proof of concept. 

Looking on `ghcide`, I think it can really benefit from using [`Data.Text.Lines`](https://hackage.haskell.org/package/text-rope-0.1/docs/Data-Text-Lines.html#t:TextLines) from `text-rope` instead of a plain `Text`, because this will improve ubiquitous line splitting. 

So I'd suggest to wait for an `lsp` release with `text-rope`-based VFS, then take another look at this issue.

**Comment by michaelpj** [comment]
Yes, definitely not planning to do anything before then. Thanks for letting me know how to do the conversion!

**Comment by Ptival** [comment]
Maybe the example in this issue makes this look too benign, but this bug makes Emacs w/ Haskell LSP completely unusable on projects that use Unicode. For instance with this file:

```haskell
{-# LANGUAGE UnicodeSyntax #-}
module Mini where
import Prelude
type 𝐿 a = [a]
```

Once you start editing the last line (e.g. try to replace `𝐿` with a normal `L` and then back), the LSP backend and what's in your buffer start becoming completely out of sync, and you get error messages such as parse errors or text that is no longer present in your buffer.

**Comment by michaelpj** [comment]
Yeah, I would expect "unusable" to be about right.

I've been surprised that there hasn't been more complaint on this issue, and I've guessed that means that Unicode just isn't that popular with Haskell developers. 

That said, I don't think this would be too hard to fix, just tedious. So if someone is keen to see it done then that could be great!

**Comment by Ptival** [comment]
I could be bothered enough to attempt a fix over the weekend, though I'm not quite sure where to begin.

Do you think the solution would look more or less like chasing the points at which positions are being exchanged between haskell-language-server and lsp, and figure out the correct invocation of your `codePointPositionToPosition` and `positionToCodePointPosition` functions, and its possible consequences on the nearby code?

**Comment by michaelpj** [comment]
Yes, that's pretty much it. I would probably use `hlint` to help: add an `error` rule for the use of the `Position` constructor directly, make a whitelist, and then start whittling it down. That way we can also hopefully stop it getting worse :sweat_smile:

**Comment by Ptival** [comment]
Almost immediately running into an odd situation. Consider:

https://github.com/haskell/haskell-language-server/blob/9565d0b2d0b7d2ddf5a982269c103b6fd0a781a0/ghcide/src/Development/IDE/GHC/Error.hs#L89-L91

My gut feeling is this is the place that is ripe for column confusion: supposedly, the `srcLocCol` is 1-based and code point-based, while `Position` is 0-based and code unit-based. However, this is nowhere near a `VirtualFile`.

Chasing up the call chain, I often end up at a `StringBuffer` as the thing looking the most like the contents of a file, e.g.:

https://github.com/haskell/haskell-language-server/blob/9565d0b2d0b7d2ddf5a982269c103b6fd0a781a0/ghcide/src/Development/IDE/Core/Preprocessor.hs#L141-L147
where the pragmas in a file are being checked and `catchSrcErrors` eventually calls `realSrcLocToPosition`, or:

https://github.com/haskell/haskell-language-server/blob/9565d0b2d0b7d2ddf5a982269c103b6fd0a781a0/ghcide/src/Development/IDE/Core/Compile.hs#L174-L189
where maybe the module contents are available in the `ms_hspp_buf` field of the `ParsedModule`...

So:

1. Should I try to grab and pass around these `StringBuffer`s all the way down? Or is this a recipe for disaster?
2. If so, should this call for versions of your helper converter functions that just take in the contents, without the extra LSP-related stuff in a `VirtualFile`?
3. And finally, these are `StringBuffer`s, not `Rope`s. What would be the expectation here?

Overall, I get the feeling that the parts of the GHCIDE code I'm looking at are fairly LSP agnostic, so reaching the gap to the `VirtualFile` abstraction feels odd.

Any thoughts?

**Comment by Bodigrim** [comment]
More trail of destruction: https://gitlab.haskell.org/ghc/ghc/-/issues/25396

**Comment by ag-eitilt** [comment]
@Ptival , did you get any further than your last post, or did the work peter out pending answers to those questions?  So far as I can tell, the main thing that would expose the majority of issues would be that `realSrcLocToPosition` you already identified -- if we change its type signature as is necessary to compute the true `Position`, that should cascade though a lot of the code.  I've not gotten the Hlint rule working yet, so one of the only two places I've found so far which would need separate handling is, ironically, the Hlint plugin, but I'm sure there's other, more subtle places.  As I mentioned in https://github.com/haskell/haskell-language-server/issues/1987#issuecomment-4366827038, `readSrcSpan` is also going to be difficult, but that one at least seems like mostly-dead code, and like there'd be a more integrated way for the `refactor` plugin to get its position info.

I've not looked into the VFS in any real detail yet, but my initial reaction is that it's just about the level of abstraction we'd want -- `Position` is itself an inherently LSP interpretation of column/char number, so tying the two together in the conversion code isn't enforcing a new restriction on scope.  (And, in fact, I'd say the code being too agnostic to the LSP data models is one of the causes of this problem.)

**Comment by Ptival** [comment]
I have almost completely paged out what I was doing out of my brain sadly.

I have a snapshot of what I started doing [here](https://github.com/Ptival/haskell-language-server/tree/vr/fix-unicode-bug), but it might take me a bit of time to figure out what I was doing there.

[It looks](https://github.com/haskell/haskell-language-server/compare/master...Ptival:haskell-language-server:vr/fix-unicode-bug) like I had tried to change the output type of `realSrcLocToPosition` to be `CodePointPosition`. It also looks like I was trying to thread a `Rope` through a lot of functions that needed it to compute conversions.

I'm not sure all of the changes I've attempted there are actually good.

**Comment by ag-eitilt** [comment]
No worries, it was a (comparatively) long time ago.  I was mostly asking to make sure I wouldn't be duplicating work you had already done -- especially if you *had* gotten a significant portion of the way through -- so those links are definitely appreciated!  Feel free to skip paging it back in if you don't want to dive back into the work; I was already planning to look at this myself, and I can read through your changes nearly as well as you can.  Or else if you did want to keep going on it, is there any way I can help?

**Comment by Ptival** [comment]
I could go either way, feel free to take over, but if you need help / someone to discuss design decisions with, I'm happy to try.

I mostly attempted this because it prevented me from working effectively with Emacs on a Unicode heavy code base, but in the mean time I just use VSCode when working on that project.

I'd love to see this fixed, but I only have some limited personal time to apply to it, especially without guidance from folks more knowledgeable with the code!
